ldap/servers/slapd/back-ldbm/db-bdb/bdb_layer.c

```diff
@@ -2586,13 +2586,6 @@ bdb_rm_db_file(backend *be, struct attrinfo *a, PRBool use_lock, int no_force_ch
     if (NULL == pEnv) { /* db does not exist */
         return rc;
     }
-    /* Added for bug 600401. Somehow the checkpoint thread deadlocked on
-     index file with this function, index file couldn't be removed on win2k.
-     Force a checkpoint here to break deadlock.
-  */
-    if (0 == no_force_checkpoint) {
-        bdb_force_checkpoint(li);
```

Comment:
I'm a bit hazy but why do you need to remove the checkpoint here?

Reply:
Short story long...
Currently when a user requests an offline reindex the CLI maps this to upgradedb, which does a reindex of each backend by doing an export and import for each backend. This is a problem for some customers as they only requested the reindex of a single backend and wanted to avoid the overhead of reindexing all backends.

When requesting an offline reindex via the server, a libdb checkpoint error occurs:
/usr/sbin/ns-slapd db2index  -D /etc/dirsrv/slapd-inst01  -n test01  -t aci

... ERR - libdb - BDB1566 txn_checkpoint interface requires an environment configured for the transaction subsystem
... ERR - bdb_force_checkpoint - Checkpoint FAILED, error Unexpected dbimpl error code (22)

The last working version of a single backend reindex was in branch 389-ds-1.4.1, so I compared the behaviour of 1.4.1 to master to understand the libdb error.

During an offline reindex the existing index file is deleted and re-generated as they server processes the specified index attributes. During the deletion of the index file a checkpoint is requested, but only if transaction logging is disabled:
https://github.com/389ds/389-ds-base/blob/aeb90eb0c41fc48541d983f323c627b2e6c328c7/ldap/servers/slapd/back-ldbm/db-bdb/bdb_layer.c#L4232-L4251

In 1.4.1 a checkpoint is never performed because transaction logging was disabled, but in master a checkpoint is attempted because transaction logging is not disabled, although it was requested:
https://github.com/389ds/389-ds-base/blob/aeb90eb0c41fc48541d983f323c627b2e6c328c7/ldap/servers/slapd/back-ldbm/db-bdb/bdb_ldif2db.c#L1409-L1412

It turns out that some attributes specific to the BDB backend are no longer in the ldbm config entry, CONFIG_DB_TRANSACTION_LOGGING is one of them.

So to disable transaction logging in libdb we should use:
https://github.com/389ds/389-ds-base/blob/aeb90eb0c41fc48541d983f323c627b2e6c328c7/ldap/servers/slapd/back-ldbm/db-bdb/bdb_ldif2db.c#L2411-L2415
Instead of:
https://github.com/389ds/389-ds-base/blob/aeb90eb0c41fc48541d983f323c627b2e6c328c7/ldap/servers/slapd/back-ldbm/db-bdb/bdb_ldif2db.c#L1411

So to answer your question, we dont need to remove the checkpoint, we just need to call the function that disabled transaction loggin in libdb. I have updated this PR to reflect this.

Reply:
Nice investigation!
A small note, I think it makes sense to bring back the removed comment too.

Reply:
It does, my bad. Reverted the comment.

Reply:
Great, thanks for the follow up investigation!